src/services/services.ts

```diff
@@ -1536,26 +1573,166 @@ namespace ts {
             return checker.getSymbolAtLocation(node);
         }
 
+
+        const sourceMapCommentRE = /^\/\/[@#] sourceMappingURL=(.+)$/gm;
+        const dataURLRE = /^data:/;
+        const base64URLRE = /^data:application\/json;charset=utf-8;base64,(.+)$/;
+        function scanForSourcemapURL(fileName: string) {
+            const mappedFile = sourcemappedFileCache.get(toPath(fileName, currentDirectory, getCanonicalFileName));
+            if (!mappedFile) {
+                return;
+            }
+            const starts = getLineStarts(mappedFile);
+            for (let index = starts.length - 1; index--; index >= 0) {
+                sourceMapCommentRE.lastIndex = starts[index];
+                const comment = sourceMapCommentRE.exec(mappedFile.text);
+                if (comment) {
+                    return comment[1];
+                }
+            }
+        }
+
+        function convertDocumentToSourceMapper(file: { sourceMapper?: sourcemaps.SourceMapper }, contents: string, mapFileName: string) {
+            let maps: sourcemaps.SourceMapData;
+            try {
+                maps = JSON.parse(contents);
+            }
+            catch {
+                // swallow error
+            }
+            if (!maps || !maps.sources || !maps.file || !maps.mappings) {
+                // obviously invalid map
+                return file.sourceMapper = sourcemaps.identitySourceMapper;
+            }
+            return file.sourceMapper = sourcemaps.decode({
+                readFile: s => host.readFile(s),
+                fileExists: s => host.fileExists(s),
+                getCanonicalFileName,
+                log,
+            }, mapFileName, maps, program, sourcemappedFileCache);
+        }
+
+        function getSourceMapper(fileName: string, file: { sourceMapper?: sourcemaps.SourceMapper }) {
+            if (!host.readFile || !host.fileExists) {
+                return file.sourceMapper = sourcemaps.identitySourceMapper;
+            }
+            if (file.sourceMapper) {
+                return file.sourceMapper;
+            }
+            let mapFileName = scanForSourcemapURL(fileName);
+            if (mapFileName && dataURLRE.exec(mapFileName)) {

```

Comment:
`dataURLRE` seems unnecessary. The same thing could be accomplished with a single RegExp.

Reply:
I'm doing this to avoid bothering to look for a file at a `data:` URL we couldn't understand. It's likely the rest of the path machinery may handle arbitrary data URLs OK-ish... but do we want it to?

Reply:
What I meant was that your regexp could be written like this `/^data:(?:application\/json;charset=utf-8;base64,(.+)$)?/`. Then you can `exec` the regexp once. If the result is non-null, its at least a data URL, but if it has a `matches[1]` then it's a valid base64 data URL.

Reply:
I don't think we want to be in the game of supporting multiple encodings, and our builtin impl (ignoring what the platform can provide) only handles utf8.

Reply:
Fair enough, though I do believe leveraging a single RegExp and possibly restricting the allowed values for the capturing group would be worthwhile.